### ISSUE TITLE
DropdownButton selected Item not Showing inside StateBuilder on Data ConnectionState

### ISSUE THREAD

**Issue by podcoder** [GIfatahTH/states_rebuilder#39]
DropdownButton `onChange` method returns the selected value, but when I am using setState() method, Selected Item does appearing on DropdownButton. Even though the value is changed.

``` dart

class TestPage extends StatefulWidget {
  TestPage({Key key}) : super(key: key);

  @override
  _TestPageState createState() => _TestPageState();
}

class _TestPageState extends State<TestPage> {
 
  Gender _allowedGender;

  @override
  Widget build(BuildContext context) {

    return Scaffold(
      appBar: AppBar(
        title: Text('Testing'),
      ),

      body: Injector(
        inject: [
          Inject<LookupsStore>(() => LookupsStore(LookupsRepositoryImpl()))
        ],
        builder: (context) {
          return StateBuilder(
            didChangeDependencies: (_, ReactiveModel<LookupsStore> model) {
              model.setState((store) => store.getLookups(),
                  onError: (_, error) {
                if (error is NetworkError) {
                  showSnackbar(context, error.message);
                } else {
                  throw error;
                }
              });
            },
            models: [Injector.getAsReactive<LookupsStore>(context: context)],
            builder: (_, reactiveModel) {
              return reactiveModel.whenConnectionState(
                  onIdle: () => loaderWidget(),
                  onWaiting: () => loaderWidget(),
                  onData: (lookupStore) => _loaded(lookupStore),
                  onError: (_) => networkErrorWidget("Network error"));
            },
          );
        },
      ),
    );
  }

  Container _loaded(LookupsStore store) {
   
    return Container(
   
      child: SingleChildScrollView(
        child: Form(
       
          child: Column(children: <Widget>[
       
                  DropdownButtonFormField<Gender>(
                      decoration: InputDecoration(
                          border: OutlineInputBorder(),
                          contentPadding: EdgeInsets.symmetric(
                              horizontal: 6)),
                      hint: Text('Gender'),
                      // isExpanded: true,
                      onChanged: (Gender val) {
                        print(_allowedGender);
                        setState(() {
                          _allowedGender = val;
                        });
                      },
                  
                      value: _allowedGender,
                      items: store.genders
                          .map(
                            (gender) => DropdownMenuItem<Gender>(
                              value: gender,
                              child: Text(gender.name),
                            ),
                          )
                          .toList(),
                    ),
                  ),
                ],
              ),
            ),
          
          ],
          );
  }
}

```

**Comment by seddiqsorush** [comment]
I am new to StateRebuiler, Kindly help me if I am doing something wrong

**Comment by GIfatahTH** [comment]
There are many ways to refactor your code. But following your implementation:

1. First try to make `TestPage` StatelessWidget.
2. Add generic type to `StateBuilder`. 
3. remove the context parameter from `Injector.getAsReactive` in the `models` parameter.
4. Use `StatefullWidget` instead of `_loaded` method.

Your code should look like this:
```dart
class TestPage extends StatelessWidget {
  TestPage({Key key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Testing'),
      ),
      body: Injector(
        inject: [
          Inject<LookupsStore>(() => LookupsStore(LookupsRepositoryImpl()))
        ],
        builder: (context) {
          //Add generic type to StateBuilder
          return StateBuilder<LookupsStore>(
            didChangeDependencies: (_, ReactiveModel<LookupsStore> model) {
              model.setState((store) => store.getLookups(),
                  onError: (_, error) {
                if (error is NetworkError) {
                  showSnackbar(context, error.message);
                } else {
                  throw error;
                }
              });
            },
            //Remove the context parameter. It is not necessary here
            models: [Injector.getAsReactive<LookupsStore>(context: context)],
            builder: (_, reactiveModel) {
              return reactiveModel.whenConnectionState(
                  onIdle: () => loaderWidget(),
                  onWaiting: () => loaderWidget(),
                  //Use StatefulWidget instead of _loaded method
                  onData: (lookupStore) => Loaded(),
                  onError: (_) => networkErrorWidget("Network error"));
            },
          );
        },
      ),
    );
  }


}

class Loaded extends StatefulWidget {
  @override
  _LoadedState createState() => _LoadedState();
}

class _LoadedState extends State<Loaded> {

  Gender _allowedGender;

  final  lookupsStoreRM = Injector.getAsReactive<LookupsStore>();
  @override
  Widget build(BuildContext context) {
    return Container(
      child: SingleChildScrollView(
        child: Form(
          child: Column(
            children: <Widget>[
              DropdownButtonFormField<Gender>(
                decoration: InputDecoration(
                    border: OutlineInputBorder(),
                    contentPadding: EdgeInsets.symmetric(horizontal: 6)),
                hint: Text('Gender'),
                // isExpanded: true,
                onChanged: (Gender val) {
                  print(_allowedGender);
                  setState(() {
                    _allowedGender = val;
                  });
                },

                value: _allowedGender,
                items: lookupsStoreRM.state.genders
                    .map(
                      (gender) => DropdownMenuItem<Gender>(
                        value: gender,
                        child: Text(gender.name),
                      ),
                    )
                    .toList(),
              ),
            ],
          ),
        ),
      ),
    );
  }
}
```

Please note that I have not tested the code. I just change it from what I theoretically expect. Please try it and let me know if it works. If it is the case, I will explain why.

**Comment by seddiqsorush** [comment]
Thank you for the quick response. Yes, it is working now, If you explain the concept it will be greate. I did use didChangeDependencies to call to a function from my `state` and then fetch the data from the server which is in the `repository`,  Initially I did try to call to the function in `didChangeDependencies` of the `StatefullWidget` itself, but I was getting in error , which states that the `LookupsStore` is called before it's been initialized.
by using didChangeDependencies the error gone.
